schema.c

```diff
@@ -741,64 +813,301 @@ _sch_load (const char *path, const char *model_list_filename)
+
+            if (result)
+            {
+                char *tmp = result;
+                result = g_strdup_printf ("%s,%s", result, split_existing[i]);
+                g_free (tmp);
+            }
+            else
+                result = g_strdup (split_existing[i]);
         }
-        copy_nsdef_to_root (instance->doc, module_new);
-        if (save_module_info (instance, module_new))
+        g_strfreev (split_existing);
+    }
+
+    if (result && g_strcmp0 (existing, result))
+    {
+        xmlSetProp (schema, (const xmlChar *)"features", (const xmlChar *) result);
+        s = (xmlNode *) sch_node_next_sibling ((sch_node *) schema);
+        while (s)
         {
-            add_module_info_to_child (instance, module_new);
-            merge_nodes (module_new->ns, module, module->children, module_new->children, 0);
-            xmlFreeDoc (item->doc_new);
-            item->doc_new = NULL;
-            assign_ns_to_root (instance->doc, module->children);
+            if (s->ns && g_strcmp0 ((char *)s->ns->href, load_model->ns_href) == 0)
+                xmlSetProp (s, (const xmlChar *)"features", (const xmlChar *) result);
+
+            s = (xmlNode *) sch_node_next_sibling ((sch_node *) s);
         }
-        else
+    }
+    else if (!result)
+    {
+        xmlSetProp (schema, (const xmlChar *)"features", (const xmlChar *) "");
+        s = (xmlNode *) sch_node_next_sibling ((sch_node *) schema);
+        while (s)
         {
-            xmlFreeDoc (item->doc_new);
-            item->doc_new = NULL;
+            if (s->ns && g_strcmp0 ((char *)s->ns->href, load_model->ns_href) == 0)
+                xmlSetProp (s, (const xmlChar *)"features", (const xmlChar *) "");
+
+            s = (xmlNode *) sch_node_next_sibling ((sch_node *) s);
         }
     }
-    g_list_free_full (files, sch_load_item_free);
 
-    /* Store a link back to the instance in the xmlDoc stucture */
-    instance->doc->_private = (void *) instance;
+    g_free (load_model->features);
+    if (result)
+        load_model->features = g_strdup (result);
+    else
+        load_model->features = NULL;
 
-    return instance;
+    yang_library_update_feature_information (load_model);
+
+    g_free (result);
+    xmlFree (existing);
+
+    if (schema)
+    {
+        existing = (char *) xmlGetProp ((xmlNode *) schema, (xmlChar *) "features");
+        xmlFree (existing);
+    }
 }
 
-sch_instance *
-sch_load (const char *path)
+static sch_loaded_model *
+sch_find_model_by_name (GList *loaded_models, const char *model_name)
 {
-    return _sch_load (path, NULL);
+    sch_loaded_model *loaded;
+    GList *list;
+
+    for (list = g_list_first (loaded_models); list; list = g_list_next (list))
+    {
+        loaded = list->data;
+        if (g_strcmp0 (model_name, loaded->model) == 0)
+            return loaded;
+    }
+    return NULL;
+}
+
+bool
+sch_update_model (sch_instance *instance, const char *model, int flags, const char *features)
```

Comment:
seems to be a bit of repeated code in here as well. Can it be made more efficient?